### PR TITLE
Add some advanced features documentation.

### LINKED ISSUE #1119: Document how to get the most speed out of Nengo
Given the cache, the optimizer (#1035), and nengo_ocl there are different options on how to run Nengo model with different implications on hardware-requirement (CPU vs GPU, RAM) and build time that influence the run of the model. We should write a summary as part of the documentation of these options.

### PR DESCRIPTION
**Pull request by jgosmann**
**Description:**
Extends the documentation with some advanced topics. In particular:
- The Nengo RC file and available settings.
- How to improve the performance.

**Motivation and context:**
The performance documentation is probably useful for new users who start to build larger models. Also it provides documentation for how to activate/deactivate the operator graph optimizer which is important because it provides a trade off between build time and simulation time.

In this documentation I had to reference the Nengo RC files which prompted me to add the missing documentation for that file and the settings too.

This PR closes #1119, addresses #1017.

**Interactions with other PRs:**
This adds documentation for the yet to be merged PR #1035. This assumes that the optimizer is active by default, but I could see this being changed. In that case this PR needs to be updated. 

**How has this been tested?**
Build the docs and skimmed over it and tested cross-reference links. Should probably get a more careful read, preferably by a native speaker.

**Where should a reviewer start?**
Build the docs and start reading the "Advanced Topics" section.

**How long should this take to review?**

- Average (neither quick nor lengthy)

**Types of changes:**

- New feature (non-breaking change which adds functionality)

**Checklist:**

- [x] I have read the **CONTRIBUTING.rst** document.
- [x] I have updated the documentation accordingly❗ 
- [x] I have included a changelog entry.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

**Still to do:**
- [x] Potential updates if adjustments are made in #1035 before it gets merged.

### CODE REVIEW

File: docs/improving_performance.rst [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+---------------
+
+*Influences build time.*
+
+A significant amount of time during the build is spent on finding the NEF
+decoders. Because of that, it is possible to cache the decoders. The first
+build of a model will still take about the same time (technically a bit longer,
+because the computed decoders will be stored), but all subsequent builds of the
+same model can just load the stored decoders and will be faster. This will also
+work to some degree if changes to the model are made, but of course only for
```
Comment (Seanny123) [inline docs/improving_performance.rst]: Remove `of course`. It's kind of condescending. Maybe people don't know what caching is in this context.

File: docs/improving_performance.rst [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+But be aware that many small ensembles will take longer to simulate if the
+operator graph optimizer (see above) is deactivated.
+
+
+Limiting probed data
+--------------------
+
+*Influences mainly memory consumption.*
+
+Obviously, all data that gets probed in the model has to be stored in memory.
```
Comment (Seanny123) [inline docs/improving_performance.rst]: Remove `obviously`. It's not needed here.

**Comment by Seanny123** [comment]
Left some inline comments to fix some phrasing.

**Comment by mgserafi** [comment]
Made minor changes to the phrasing. Added @Seanny123's suggested changes.

**Comment by tbekolay** [comment]
I also took a look at this, only partly in order to get my review in for the week ;)

Docs look great, really nice addition, and nice to have an advanced features section for more esoteric information :+1: I pushed a commit with some slight rewordings. I also added the WIP label because this is blocked by #1035.

One thing that I would like to do (but am not sure how to do best) is not duplicate the `nengorc` information. Right now, that information is nicely laid out in the new `nengorc.rst` doc, but similar (yet not the same) documentation exists in `nengo-data/nengorc`, and also in `nengo/rc.py`. Ideally it would only be in one place, so the question is where. I think I would prefer it to be in the new doc that's in this PR, and then we should link that that document (somehow) in `rc.py`, and remove it from `nengorc`.

One thing that [matplotlib does](http://matplotlib.org/users/customizing.html) is link to and include a copy of a sample rc file in there, which we could also do with the [`include` directive](http://docutils.sourceforge.net/docs/ref/rst/directives.html#include). I think it'd be especially nice if the `nengorc` included there is barebones instead of the super verbose version that matplotlib has (since the `.rst` file will always be able to render a nicer version of the docs).

**Comment by jgosmann** [comment]
> I think it'd be especially nice if the nengorc included there is barebones instead of the super verbose version that matplotlib has

I think it is nice to have a verbose sample rc file because then I can edit it and have all the required information right there without the need to navigate to the documentation.

**Comment by tbekolay** [comment]
'kay, then why don't we include the verbose sample in the docs rather than duplicating the information?

**Comment by jgosmann** [comment]
Sure, but the sample file will need some updating (probably should've been updated in the first place).

**Comment by tbekolay** [comment]
Pushed a commit updating the sample file and `rc.py` so that there's only a tiny bit of duplication of information. In the generated output, the RC settings docs look pretty good to me now (the performance doc already looked good).

If you want to see the generated docs, I can send put the HTML files up somewhere, or you can try building the docs yourself. I do `python setup.py build_sphinx`; the only oddity is that you currently have to downgrade to `ipython[all]==2.4.1` as the doc building hasn't been updated; I was working on updating it in the [`nbsphinx` branch](https://github.com/nengo/nengo/tree/nbsphinx) but that hit a few issues I haven't solved yet.

**Comment by jgosmann** [comment]
> python setup.py build_sphinx

`error: invalid command 'build_sphinx'`

**Comment by jgosmann** [comment]
(I didn't have the requirement installed, but the error message is not really helpful in that case.)

**Comment by tbekolay** [comment]
> (I didn't have the requirement installed, but the error message is not really helpful in that case.)

Yeah... we could add something to `setup.py` to detect that but it doesn't really seem worth the additional code as probably three people in the world have tried to build the docs.

**Comment by jgosmann** [comment]
To dos:
- [ ] Mention that the optimizer can do certain optimizations only if SciPy is installed.
- [ ] Add configuration options for optimizer to sample rc file.

**Comment by tbekolay** [comment]
> Add configuration options for optimizer to sample rc file.

I did in my commit, I think... at least, I added the ones that were in the `nengorc.rst` doc.

**Comment by jgosmann** [comment]
Seem to be missing to me ... (or maybe I forgot to add those in the first place)

**Comment by jgosmann** [comment]
The `[OpMergeOptimizer]` section heading is missing (that was what I was searching for).

**Comment by tbekolay** [comment]
I added two options, see https://github.com/nengo/nengo/pull/1130/files#diff-e6c6a36ed8df7b1bd3ac3887920554ccR60, not sure if there are more...

**Comment by tbekolay** [comment]
Ah, right you are! Should probably be `op_merge_optimizer` to match the other sections, but yeah, I missed that.

**Comment by jgosmann** [comment]
> Should probably be op_merge_optimizer to match the other sections, but yeah, I missed that.

It needs to match the class name with the current implementation.

**Comment by jgosmann** [comment]
Fixed those things.

**Comment by tbekolay** [comment]
As per the discussion at the dev meeting, when this is merged, it should close #340.

**Comment by jgosmann** [comment]
Added commits to update this to match what has actually been merged to master.

**Comment by Seanny123** [comment]
This is still good. The `flake8` TravisCI errors should be fixed by bringing this branch up to date and I'm sure whoever resolves the conflicts will fix `CHANGES.rst` to refer to the right Nengo version.

**Comment by jgosmann** [comment]
LGTM 🍰

**Pull request merged**